SoundCloud Playlist Export & Import functionalities

rimma-kubanova opened the pull request:
Implemented export and import features for SoundCloud

IMPORT:
* Added modal for SoundCloudPlaylistImport 
* Added endpoints for SoundCloud Playlist import and playlist tracks import

EXPORT:
* Added troi call for SoundCloud export
* Added endpoint for SoundCloud export on back
* Added button for SoundCloud export on front

pep8speaks commented:
Hello @rimma-kubanova! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`listenbrainz/metadata_cache/soundcloud/client.py`](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/metadata_cache/soundcloud/client.py):

> [Line 9:15](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/metadata_cache/soundcloud/client.py#L9): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 11:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/metadata_cache/soundcloud/client.py#L11): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`listenbrainz/troi/export.py`](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/troi/export.py):

> [Line 28:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/troi/export.py#L28): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`listenbrainz/troi/import_ms.py`](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/troi/import_ms.py):

> [Line 40:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/troi/import_ms.py#L40): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`listenbrainz/webserver/views/playlist_api.py`](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py):

> [Line 877:15](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py#L877): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 908:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py#L908): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 914:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py#L914): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 916:131](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py#L916): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (133 > 130 characters)
> [Line 926:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py#L926): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 929:1](https://github.com/metabrainz/listenbrainz-server/blob/78f62c8d78b379f83ef42bd8bad0a30b81cf8900/listenbrainz/webserver/views/playlist_api.py#L929): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

anshg1214 commented on frontend/js/src/user/playlists/components/ImportSoundCloudPlaylistModal.tsx:
```diff
@@ -0,0 +1,203 @@
+          <ToastMsg
+            title="Error loading playlists"
+            message={
+              error?.message === "Forbidden" ? (
+                <>
+                  Session has expired. Please reconnect to{" "}
+                  <Link to="/settings/music-services/details/">Soundcloud</Link>
+                  .
+                </>
+              ) : (
+                error?.message ?? error
+              )
```
Instead of handling error message `Forbidden`, you should handle the error `Unauthorized` and `Not Found`. [link](https://developers.soundcloud.com/docs/api/explorer/open-api#/me/get_me_playlists)

anshg1214 reviewed (changes requested):
The PR looks great! 

I'm not able to import the playlist from soundcloud. The error message says 
```
UnboundLocalError: cannot access local variable 'name' where it is not associated with a value
```
![image](https://github.com/user-attachments/assets/2c93b5ac-e189-42cd-8945-d36bcf732c0d)

rimma-kubanova commented:
> The PR looks great!
> 
> I'm not able to import the playlist from soundcloud. The error message says
> 
> ```
> UnboundLocalError: cannot access local variable 'name' where it is not associated with a value
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/84081628/363161394-2c93b5ac-e189-42cd-8945-d36bcf732c0d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNzAzOTMsIm5iZiI6MTcyNTI3MDA5MywicGF0aCI6Ii84NDA4MTYyOC8zNjMxNjEzOTQtMmM5M2I1YWMtZTE4OS00MmNkLTg5NDUtZDM2YmNmNzMyYzBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDA5NDEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1ZTM4NWNjNmRhN2U2NDJjZmVmYTgyZjRiZWY3MWI1Y2NlYmEwYTMzOTJhYjUxZjJjZDM4NzVjZTU0NDEzNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rd-oAg1jSNSoCmzwid-68z0Xim4b01khqb9uWQLbXpo)

Hi @anshg1214, thanks for catching that! It looks like I updated the code for export but overlooked the import adjustments. 😅 I’ve fixed it now in [this PR](https://github.com/metabrainz/troi-recommendation-playground/pull/150).

I'm also working on the other requested changes!

MonkeyDo commented on listenbrainz/webserver/views/playlist_api.py:
```diff
@@ -918,21 +929,24 @@ def import_playlist_from_music_service(service):
         raise APIBadRequest(f"Service {service} is not linked. Please link your {service} account first.")
 
-    if service == "apple_music" and not token["refresh_token"]:
+    if service == "apple_music" and not token.get("refresh_token"):
```
I want to flag that this token check is not present in some of the other functions where the spotify token check that follows does appear.

Should the same check be added?
If so I will do some more refactoring to avoid duplicating the code again, and make a separate function to check the token.

MonkeyDo commented:
@amCap1712 As discussed, the PR is now up to date with master.

Please do have another pass checking the python files as I had to manually backport some changes, mainly in playlist_api.py (the merge conflict was messy so I discarded the changes and implemented them back in).|

Also please see the question above regarding apple music refresh_token checking.

MonkeyDo commented:
I put the PR on test.listenbrainz.org.
Looks like I am able to import a playlist from SC, but can't seem to be able to export from LB to SC.

There are issues with the Apple Music exporter as well, so this could potentially be a wider Troi patch problem.

For reference, on the front-end the only error message I get is "Error: 'NoneType' object is not subscriptable`
On the server side the only output doesn't show an error, but does have a "None" value where a playlist URL should be (logical if the playlist generation failed, but should probably catch that an throw a more helpful error)
```
Troi playlist generation starting...
  PlaylistFromJSPFElement                   1 items
  PlaylistElement                                    1 items
done.
Submitted playlist to soundcloud: None

playlist: 'My frist playlist'
Night or Day                                                 Franz Ferdinand                                    7b506 aa7a2               
A playlist with 1 tracks was generated.

```

amCap1712 commented:
@MonkeyDo Is it fine to merge this PR? The export issues are because of the soundcloud data and I will resolve them separately.

MonkeyDo commented:
> @MonkeyDo Is it fine to merge this PR? The export issues are because of the soundcloud data and I will resolve them separately.

OK for me,but there are failing tests. Is that part of the issues in question?
